kong/utils.go

```diff
@@ -141,3 +143,87 @@ func VersionFromInfo(info map[string]interface{}) string {
 	}
 	return version.(string)
 }
+
+func getDefaultProtocols(schema gjson.Result) []*string {
+	var res []*string
+	fields := schema.Get("fields")
+
+	for _, field := range fields.Array() {
+		protocols := field.Map()["protocols"]
+		d := protocols.Get("default")
+		if d.Exists() {
+			for _, v := range d.Array() {
+				res = append(res, String(v.String()))
+			}
+			return res
+		}
+	}
+	return res
+}
+
+func fillConfigRecord(schema gjson.Result, config Configuration) Configuration {
+	res := config.DeepCopy()
+	value := schema.Get("fields")
+
+	value.ForEach(func(key, value gjson.Result) bool {
+		// get the key name
+		ms := value.Map()
+		fname := ""
+		for k := range ms {
+			fname = k
+			break
+		}
+
+		if fname == "config" {
+			newConfig := fillConfigRecord(value.Get(fname), config)
+			res = newConfig
+			return true
+		}
+
+		// check if key is already set in the config
+		if _, ok := config[fname]; ok {
+			// yes, don't set it
+			return true
+		}
+		ftype := value.Get(fname + ".type")
+		if ftype.String() == "record" {
+			subConfig := config[fname]
+			if subConfig == nil {
+				subConfig = make(map[string]interface{})
+			}
+			newSubConfig := fillConfigRecord(value.Get(fname), subConfig.(map[string]interface{}))
+			res[fname] = map[string]interface{}(newSubConfig)
+			return true
+		}
+		value = value.Get(fname + ".default")
+		if value.Exists() {
+			res[fname] = value.Value()
+		} else {
+			// if no default exists, set an explicit nil
+			res[fname] = nil
+		}
+		return true
+	})
+
+	return res
+}
+
+// FillPluginsDefaults ingests plugin's defaults from its schema
+func FillPluginsDefaults(plugin *Plugin, schema map[string]interface{}) (*Plugin, error) {
```

Comment:
@GGabriele I made a mistake in the review.
This method interface is misleading and wrong.
The code takes in a plugin struct and then mutates it.
But the method signature also returns a plugin resource.

Either of the following are fine (but together not):
- Return only an error, and document that the plugin is filled and mutated
- Inside the function, make a copy of the input plugin, mutate the copy and return it

Reply:
Doh, my bad, sorry :(

https://github.com/Kong/go-kong/pull/116